### PR TITLE
Allow to start the Quartz scheduler in halted mode

### PR DESCRIPTION
**Pull request by gsmet**
Supersedes https://github.com/quarkusio/quarkus/pull/14043

/cc @nragon @machi1990

### CODE REVIEW

**Review by machi1990 — APPROVED** [review]

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * Test suite for deploying a job while quartz scheduler in halted state
+ */
+public final class HaltedScheduleJobTest {
+
+    @SuppressWarnings("unused")
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(Starter.class)
+                    .addAsResource(new StringAsset("quarkus.quartz.enabled=true\n" +
```
Comment (mkouba) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java:37]: I believe that there is no such config property like `quarkus.quartz.enabled`...

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                            "quarkus.quartz.start-mode=halted"),
+                            "application.properties"));
+    @Inject
+    Instance<org.quartz.Scheduler> quartzScheduler;
```
Comment (mkouba) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java:41]: Is there a reason to use `javax.enterprise.inject.Instance` here?

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * Asserts if a job was executed or not
+     *
+     * @param expectedState expected state for the job execution
+     */
+    private void assertJobExecution(final boolean expectedState) throws InterruptedException {
+
+        assertEquals(expectedState, Starter.LATCH.await(3, TimeUnit.SECONDS),
```
Comment (mkouba) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/HaltedScheduleJobTest.java:75]: Hm, so this test will always take at least 3 seconds... Maybe `HaltedSchedulerTest` that asserts the return value of `org.quartz.Scheduler.isInStandbyMode()` is enough and we could remove this test completely?

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java [hunk 4]
```diff
@@ -65,25 +65,28 @@
...
+        final QuartzRuntimeConfig runtimeConfig = quartzSupport.getRuntimeConfig();
+        warnDeprecated(runtimeConfig);
+
+        boolean forceStart;
+        if (runtimeConfig.startMode != QuartzStartMode.NORMAL) {
+            startHalted = (runtimeConfig.startMode == QuartzStartMode.HALTED);
+            forceStart = startHalted || (runtimeConfig.startMode == QuartzStartMode.FORCED);
+        } else {
+            startHalted = false;
+            forceStart = runtimeConfig.forceStart.isPresent() && runtimeConfig.forceStart.get();
```
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:83]: `runtimeConfig.forceStart.orElse(false)`?

**Review by mkouba — CHANGES_REQUESTED** [review]

**Comment by gsmet** [comment]
@mkouba I applied your suggestions and renamed `halted` to `standby`. Should be good to go now.

**Review by mkouba — APPROVED** [review]

**Comment by nragon** [comment]
@gsmet not to be confused with quartz standby mode (which can affect behavior on resuming a trigger)

**Comment by gsmet** [comment]
@nragon so you're saying renaming it to `standby` is a bad idea?

From my tests, the Quartz scheduler was in standby mode on startup.

**Comment by mkouba** [comment]
> @gsmet (note) not to be confused with quartz standby mode (which can affect behavior on resuming a trigger)

Hm, that's actually a good point. So `standby` means started but temporarily halted.

By the way, the current `io.quarkus.quartz.runtime.QuartzScheduler.isRunning()` implementation does not reflect the new start mode because previously an enabled scheduler was always started.

**Comment by mkouba** [comment]
> @nragon so you're saying renaming it to `standby` is a bad idea?
> 
> From my tests, the Quartz scheduler was in standby mode on startup.

It's paused by default: https://github.com/quartz-scheduler/quartz/blob/master/quartz-core/src/main/java/org/quartz/core/QuartzSchedulerThread.java#L123

**Comment by gsmet** [comment]
> By the way, the current io.quarkus.quartz.runtime.QuartzScheduler.isRunning() implementation does not reflect the new start mode because previously an enabled scheduler was always started.

Not sure I understand that? `isRunning()` returns false if we set the start mode to `halted`.

**Comment by gsmet** [comment]
Got back to `halted` then.

**Comment by mkouba** [comment]
> > By the way, the current io.quarkus.quartz.runtime.QuartzScheduler.isRunning() implementation does not reflect the new start mode because previously an enabled scheduler was always started.
> 
> Not sure I understand that? `isRunning()` returns false if we set the start mode to `halted`.

Yeah, it works because `isInStandbyMode()` returns true by default - see https://github.com/quarkusio/quarkus/pull/14122#issuecomment-755207595

**Comment by nragon** [comment]
I was just mentioning that the standby action on quartz implies notifying both job store and listeners.
For me halted means that scheduler was not put on hold but the start was delayed for some reason.

**Comment by gsmet** [comment]
@nragon yes, makes sense. I went back to your original proposal, sorry for the noise.

**Pull request merged**